openmc/mesh.py

```diff
@@ -217,22 +275,177 @@ def write_data_to_vtk(self, points, filename, datasets, volume_normalization=Tru
         volume_normalization : bool, optional
             Whether or not to normalize the data by
             the volume of the mesh elements.
+        curvilinear : bool
+            Whether or not to write curvilinear elements. Only applies to
+            ``SphericalMesh`` and ``CylindricalMesh``.
 
         Raises
         ------
         ValueError
             When the size of a dataset doesn't match the number of mesh cells
 
+        Returns
+        -------
+        vtk.StructuredGrid or vtk.UnstructuredGrid
+            a VTK grid object representing the mesh
+        """
+        import vtk
+
+        # check that the data sets are appropriately sized
+        if datasets is not None:
+            self._check_vtk_datasets(datasets)
+
+        # write linear elements using a structured grid
+        if not curvilinear or isinstance(self, (RegularMesh, RectilinearMesh)):
+            vtk_grid = self._create_vtk_structured_grid()
+            writer = vtk.vtkStructuredGridWriter()
+        # write curvilinear elements using an unstructured grid
+        else:
+            vtk_grid = self._create_vtk_unstructured_grid()
+            writer = vtk.vtkUnstructuredGridWriter()
+
+        # maintain a list of the datasets as added
+        # to the VTK arrays to ensure they persist
+        # in memory until the file is written
+        if datasets is not None:
+            datasets_out = self._apply_vtk_datasets(vtk_grid, datasets, volume_normalization)
```

Comment:
Return value is not used -- is that intentional?

Reply:
Yep. The comment above explains why. If we don't keep the `datasets_out` object that holds the NumPy dataset arrays in memory then VTK will write junk data to the file b/c the VTK arrays in `_apply_vtk_datasets` adopts the memory address of the NumPy arrays (this is not documented in the VTK Python interface to my knowledge btw).

Maybe there's a better way to handle this...

Reply:
I opted to remove the `_apply_vtk_datasets` function. I think it's muddying the water here regarding why datasets_out is returned and held in the `write_data_to_vtk` method.